src/ring_theory/non_unital_subsemiring/basic.lean

```diff
@@ -0,0 +1,814 @@
+  s.comap ((g : S →ₙ+* T).comp (f : R →ₙ+* S)) :=
+rfl
+
+/-- The image of a non-unital subsemiring along a ring homomorphism is a non-unital subsemiring. -/
+def map (f : F) (s : non_unital_subsemiring R) : non_unital_subsemiring S :=
+{ carrier := f '' s,
+  .. s.to_subsemigroup.map (f : R →ₙ* S), .. s.to_add_submonoid.map (f : R →+ S) }
+
+@[simp] lemma coe_map (f : F) (s : non_unital_subsemiring R) : (s.map f : set S) = f '' s := rfl
+
+@[simp] lemma mem_map {f : F} {s : non_unital_subsemiring R} {y : S} :
+  y ∈ s.map f ↔ ∃ x ∈ s, f x = y :=
+set.mem_image_iff_bex
+
+@[simp] lemma map_id : s.map (non_unital_ring_hom.id R) = s :=
+set_like.coe_injective $ set.image_id _
+
+-- unavoidable coercions?
+lemma map_map (g : G) (f : F) :
+  (s.map (f : R →ₙ+* S)).map (g : S →ₙ+* T) = s.map ((g : S →ₙ+* T).comp (f : R →ₙ+* S)) :=
+set_like.coe_injective $ set.image_image _ _ _
+
+lemma map_le_iff_le_comap {f : F} {s : non_unital_subsemiring R} {t : non_unital_subsemiring S} :
+  s.map f ≤ t ↔ s ≤ t.comap f :=
+set.image_subset_iff
+
+lemma gc_map_comap (f : F) :
+  @galois_connection (non_unital_subsemiring R) (non_unital_subsemiring S) _ _ (map f) (comap f) :=
+λ S T, map_le_iff_le_comap
+/-- A non-unital subsemiring is isomorphic to its image under an injective function -/
+
+noncomputable def equiv_map_of_injective
+  (f : F) (hf : function.injective (f : R → S)) : s ≃+* s.map f :=
+{ map_mul' := λ _ _, subtype.ext (map_mul f _ _),
+  map_add' := λ _ _, subtype.ext (map_add f _ _),
+  ..equiv.set.image f s hf }
+
+@[simp] lemma coe_equiv_map_of_injective_apply
+  (f : F) (hf : function.injective f) (x : s) :
+  (equiv_map_of_injective s f hf x : S) = f x := rfl
+
+end non_unital_subsemiring
+
+namespace non_unital_ring_hom
+
+variables {F G : Type*} [non_unital_ring_hom_class F R S] [non_unital_ring_hom_class G S T]
+  (f : F) (g : G)
+
+/-- The range of a non-unital ring homomorphism is a non-unital subsemiring.
+See note [range copy pattern]. -/
+def srange : non_unital_subsemiring S :=
+((⊤ : non_unital_subsemiring R).map (f : R →ₙ+* S)).copy (set.range f) set.image_univ.symm
+
+@[simp] lemma coe_srange : (@srange R S _ _ _ _ f : set S) = set.range f := rfl
+
+@[simp] lemma mem_srange {f : F} {y : S} : y ∈ (@srange R S _ _ _ _ f) ↔ ∃ x, f x = y :=
+iff.rfl
+
+lemma srange_eq_map : @srange R S _ _ _ _ f = (⊤ : non_unital_subsemiring R).map f :=
+by { ext, simp }
+
+lemma mem_srange_self (f : F) (x : R) : f x ∈ @srange R S _ _ _ _ f :=
+mem_srange.mpr ⟨x, rfl⟩
+
+lemma map_srange : (srange f).map g = ((g : S →ₙ+* T).comp (f : R →ₙ+* S)).srange :=
+by simpa only [srange_eq_map] using (⊤ : non_unital_subsemiring R).map_map g f
+
+/-- The range of a morphism of non-unital semirings is a fintype, if the domain is a fintype.
+Note: this instance can form a diamond with `subtype.fintype` in the
+  presence of `fintype S`.-/
+instance fintype_srange [fintype R] [decidable_eq S] (f : F) : fintype (srange f) :=
```

Comment:
How about using `finite` instead of `fintype`?
```suggestion
instance finite_srange [finite R] (f : F) : finite (@srange _ S _ _ _ _ f) :=
```
As a bonus, the comment about a diamond can then be dropped since `finite` has no data.